### DISCUSSION TITLE
Tips on golang setup in lazyvim

### DISCUSSION THREAD

**Discussion by bjartek** [LazyVim/LazyVim#682] [Q&A]
This is my current configuration based on a barely naked lazyvim

https://github.com/bjartek/dotfiles/blob/master/.config/lazyvim/lua/plugins/go.lua

Witht his setup I do not get goimports on save, and I would really like that. Can anybody assist. And if you have any other tips to make this even more awesome let me know. 

original config based of https://github.com/LazyVim/LazyVim/discussions/503.

**Comment by dpetka2001** [comment]
@amaanq had posted in another topic his configuration about golang. Maybe you could take some parts from [his](https://github.com/amaanq/nvim-config/blob/master/lua/plugins/extras/lang/go.lua) to incorporate to yours?

**Reply by bjartek** [reply]
I see that you use _jay-babu/mason-null-ls.nvim_ @amaanq while in my config i use _jose-elias-alvarez/null-ls.nvim_ i have to dig into the difference here.

**Reply by Jomik** [reply]
`mason-null-ls` is the mason integration with `null-ls`. so `mason-null-ls` can automatically install programs to use with `null-ls`.
I set up my `null-ls` dependencies here https://github.com/Jomik/dotfiles/blob/718c8f82963dfc0f0a1c4c25725b75223024a68c/.config/nvim/lua/plugins/lang/go.lua#L25-L33
and configure `mason-null-ls` to install them automatically here https://github.com/Jomik/dotfiles/blob/718c8f82963dfc0f0a1c4c25725b75223024a68c/.config/nvim/lua/plugins/lsp.lua#L2-L8

**Comment by bjartek** [comment]
Going to take a look at that @dpetka2001 .

 I am just wondering why is the syntax in his config different from the one I use and the one I have seen in videos for lazyvim. What is the recommended way of doing things?

```
return {
  "nvim-treesitter/nvim-treesitter",
  opts = function(_, opts)
    vim.list_extend(opts.ensure_installed, {
      "go",
      "gomod",
      "gowork",
    })
  end,
}
```

vs

```
local util = require("util")

return {

  -- Add Go & related to treesitter
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      if type(opts.ensure_installed) == "table" then
        util.list_insert_unique(opts.ensure_installed, { "go", "gomod", "gosum", "gowork" })
      end
    end,
  },
```

**Reply by dpetka2001** [reply]
I don't believe there is a recommended way. The second way uses `list_insert_unique` which is a function he must have implemented himself for extending list tables with unique entries in Lua (probably declared in `util`, hence the `require('util')` on the top). If you are proficient at Lua you can do all sort kind of things. I am not however and prefer to stick to the way I read from the docs. You should just take ideas about his setup and not worry about syntax. You can still use the syntax from the docs to implement the setup. And if you get proficient in the future, you may also incorporate your own functions for doing things.

**Reply by amaanq** [reply]
duplicate items in nvim-ts will crash on install, hence the list_insert_unique function

**Reply by bjartek** [reply]
that sounds like a bug that should be fixed. thanks for the heads up @amaanq.

**Comment by bjartek** [comment]
Update: format on save works, it is imports on save i really want to have fixed.

**Reply by Jomik** [reply]
Did you enable `goimports` in null-ls? https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTINS.md#goimports

**Reply by bjartek** [reply]
yes

**Comment by Jomik** [comment]
Here is mine, for reference https://github.com/Jomik/dotfiles/blob/718c8f82963dfc0f0a1c4c25725b75223024a68c/.config/nvim/lua/plugins/lang/go.lua

**Reply by bjartek** [reply]
i see that yours does not have goimports but gofumpt, i assume that does the same and more?

**Comment by bjartek** [comment]
I actually think my problem is something different then what i Assumed. I want to import missing imports on save. So kind of run the lsp action Organize improts on save. So I do not think it is related to null-ls at all

**Reply by bjartek** [reply]
I added something like this and now it works
https://github.com/neovim/nvim-lspconfig/issues/115#issuecomment-1130373799

**Comment by cbrake** [comment]
In digging around today, I found the following:

https://github.com/mvdan/gofumpt#frequently-asked-questions

> > How can I use gofumpt if I already use goimports to replace gofmt?
> 
> Most editors have replaced the goimports program with the same functionality provided by a language server like gopls. This mechanism is significantly faster and more powerful, since the language server has more information that is kept up to date, necessary to add missing imports.
> 
> As such, the general recommendation is to let your editor fix your imports - either via gopls, such as VSCode or vim-go, or via their own custom implementation, such as GoLand. Then follow the install instructions above to enable the use of gofumpt instead of gofmt.

https://github.com/golang/tools/blob/master/gopls/doc/vim.md#imports

> Use the following configuration to have your imports organized on save using the logic of goimports. Note: this requires Neovim v0.7.0 or later.
> 
> ```
> vim.api.nvim_create_autocmd('BufWritePre', {
>   pattern = '*.go',
>   callback = function()
>     vim.lsp.buf.code_action({ context = { only = { 'source.organizeImports' } }, apply = true })
>   end
> })
> ```

So after installing gopls via Mason, and then adding the above to `lazy.lua`, imports are getting fixed up on save without using goimports.

**Reply by Jomik** [reply]
Pretty certain that ends up leaving your buffer in a `modified` state, which is probably not what you want, since you just saved it.

**Reply by cbrake** [reply]
Looks like you are correct -- two saves are required :-(

So I wonder what the correct solution is to get neovim to use gopls to manage imports ...

**Reply by Nikola-Milovic** [reply]
Did you get this working in the end? Can you share your config for `goimports` to work without leaving it in modified state?

**Reply by brunsgaard** [reply]
I am also interested in this, 

I use an autocmd in `lua/config/autocmds.lua`

```
vim.api.nvim_create_autocmd("BufWritePre", {
  pattern = "*.go",
  callback = function()
    vim.lsp.buf.code_action({
      context = { only = { "source.organizeImports" } },
      apply = true,
    })
  end,
})
``` 

This seems to work fine, I am unsure if this is the recommended way, and also I get a notify box on every save if there is no missing imports. 
<img width="781" alt="image" src="https://github.com/LazyVim/LazyVim/assets/815816/d59a52dd-ed7a-4ca7-a6c2-0e489be45f84">
Please share if a fix have been found :)

**Reply by cbrake** [reply]
I ended up with:

```
vim.api.nvim_create_autocmd("BufWritePre", {
  pattern = "*.go",
  callback = function()
    local params = vim.lsp.util.make_range_params()
    params.context = { only = { "source.organizeImports" } }
    -- buf_request_sync defaults to a 1000ms timeout. Depending on your
    -- machine and codebase, you may want longer. Add an additional
    -- argument after params if you find that you have to write the file
    -- twice for changes to be saved.
    -- E.g., vim.lsp.buf_request_sync(0, "textDocument/codeAction", params, 3000)
    local result = vim.lsp.buf_request_sync(0, "textDocument/codeAction", params)
    for cid, res in pairs(result or {}) do
      for _, r in pairs(res.result or {}) do
        if r.edit then
          local enc = (vim.lsp.get_client_by_id(cid) or {}).offset_encoding or "utf-16"
          vim.lsp.util.apply_workspace_edit(r.edit, enc)
        end
      end
    end
    vim.lsp.buf.format({ async = false })
  end
})
```

https://github.com/golang/tools/blob/master/gopls/doc/vim.md#neovim-imports has been updated, so that is where I got it.

Before I had problems with the process of adding imports messing up code and doing strange things. Now it seems much more stable.

**Comment by tolik505** [comment]
This setup worked for me
```
return {
  "stevearc/conform.nvim",
  dependencies = { "mason.nvim" },
  lazy = true,
  cmd = "ConformInfo",
  opts = {
    formatters_by_ft = {
      javascript = { "prettier" },
      typescript = { "prettier" },
      json = { "prettier" },
      yaml = { "prettier" },
      markdown = { "prettier" },
      graphql = { "prettier" },
      lua = { "stylua" },
      go = { "goimports", "gofumpt" },
    },
    formatters = {
      golines = {
        args = { "-m", "120" },
      },
    },
    format_on_save = {
      lsp_fallback = true,
      async = false,
      timeout_ms = 2000,
    },
  },
}
```
